File: tests/test_llm_chaos.py
import pytest
import asyncio
import aiohttp
import random
import string
import os
import sys
import psutil


# Add project root to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from config_system import config


class TestLLMChaos:
    """Explicitly verify 'Iron-Clad' resilience of the LLM Engine."""

    @pytest.mark.asyncio
    async def test_concurrency_flood(self):
        """Flood the engine with 50 concurrent requests."""
        if not config.BIN_DIR.exists():
            pytest.skip("No engine binary found")

        async def send_request(session, i):
            """Send request."""
            try:
                payload = {"prompt": f"Test request {i}", "n_predict": 10}
                async with session.post(
                    f"http://127.0.0.1:{config.llm_port}/completion", json=payload, timeout=5
                ) as resp:
                    return resp.status
            except Exception as e:
                return f"Error: {e}"

        async with aiohttp.ClientSession() as session:
            tasks = [send_request(session, i) for i in range(50)]
            results = await asyncio.gather(*tasks)

            # We expect 200s or 503s (Busy), but NOT crashes (Connection Refused)
            # If server crashes, it's a fail.
            crashes = [r for r in results if isinstance(r, str) and "Connection refused" in r]
            assert len(crashes) == 0, f"Engine crashed under load: {crashes[:3]}"

    @pytest.mark.asyncio
    async def test_tokenizer_stress(self):
        """Send a massive 1MB payload to stress the tokenizer."""
        # Generating 1MB of random garbage text
        garbage = "".join(random.choices(string.ascii_letters + string.punctuation, k=1024 * 1024))

        payload = {"prompt": garbage, "n_predict": 1}

        async with aiohttp.ClientSession() as session:
            try:
                # Should return 400 (Context Limit) or handle gracefully
                async with session.post(
                    f"http://127.0.0.1:{config.llm_port}/completion", json=payload, timeout=10
                ) as resp:
                    assert resp.status in [200, 400, 413, 500], f"Unexpected status: {resp.status}"
            except Exception as e:
                # Timeout is acceptable for massive payload, but crash is not
                if "Connection refused" in str(e):
                    pytest.fail("Engine crashed on massive payload!")

    def test_recovery_logic(self):
        """Identify Llama Server process, KILL IT, and verify it comes back."""
        import time
        from utils import is_port_active

        # 1. Verify it's running originally
        assert is_port_active(config.llm_port), "Engine not running at start of test"

        # 2. Find the PID
        target_pid = None
        for p in psutil.process_iter(["pid", "name", "cmdline"]):
            if "llama-server" not in p.info["name"].lower():
                continue
            target_pid = p.info["pid"]
            break

        if not target_pid:
            pytest.skip("Could not find llama-server process to kill")

        # 3. KILL IT
        # [X-Ray auto-fix] print(f"\n💀 KILLING PID {target_pid} FOR SCIENCE...")
        psutil.Process(target_pid).kill()

        # 4. Wait for Auto-Recovery (Max 15s)
        # The Orchestrator loop runs every ~2s
        recovered = False
        for _ in range(15):
            time.sleep(1)
            if is_port_active(config.llm_port):
                recovered = True
                break

        assert recovered, "Orchestrator FAILED to restart the engine after crash!"
        print("\n✅ Engine successfully resurrected! 🧟")

File: rust_output/src/test_llm_chaos.rs
use anyhow::{Result, Context};
use crate::config_system::{config};
use std::collections::HashMap;
use tokio;

/// Explicitly verify 'Iron-Clad' resilience of the LLM Engine.
#[derive(Debug, Clone)]
pub struct TestLLMChaos {
}

impl TestLLMChaos {
    /// Flood the engine with 50 concurrent requests.
    pub async fn test_concurrency_flood(&self) -> Result<()> {
        // Flood the engine with 50 concurrent requests.
        if !config::BIN_DIR.exists() {
            pytest.skip("No engine binary found".to_string());
        }
        let send_request = |session, i| {
            // Send request.
            // try:
            {
                let mut payload = HashMap::from([("prompt".to_string(), format!("Test request {}", i)), ("n_predict".to_string(), 10)]);
                let mut resp = session.post(format!("http://127.0.0.1:{}/completion", config::llm_port), /* json= */ payload, /* timeout= */ 5);
                {
                    resp.status
                }
            }
            // except Exception as e:
        };
        let mut session = aiohttp.ClientSession();
        {
            let mut tasks = 0..50.iter().map(|i| send_request(session, i)).collect::<Vec<_>>();
            let mut results = asyncio.gather(/* *tasks */).await;
            let mut crashes = results.iter().filter(|r| (/* /* isinstance(r, str) */ */ true && r.contains(&"Connection refused".to_string()))).map(|r| r).collect::<Vec<_>>();
            assert!(crashes.len() == 0, "Engine crashed under load: {}", crashes[..3]);
        }
    }
    /// Send a massive 1MB payload to stress the tokenizer.
    pub async fn test_tokenizer_stress(&self) -> Result<()> {
        // Send a massive 1MB payload to stress the tokenizer.
        let mut garbage = random.choices((string.ascii_letters + string.punctuation), /* k= */ (1024 * 1024)).join(&"".to_string());
        let mut payload = HashMap::from([("prompt".to_string(), garbage), ("n_predict".to_string(), 1)]);
        let mut session = aiohttp.ClientSession();
        {
            // try:
            {
                let mut resp = session.post(format!("http://127.0.0.1:{}/completion", config::llm_port), /* json= */ payload, /* timeout= */ 10);
                {
                    assert!(vec![200, 400, 413, 500].contains(&resp.status), "Unexpected status: {}", resp.status);
                }
            }
            // except Exception as e:
        }
    }
    /// Identify Llama Server process, KILL IT, and verify it comes back.
    pub fn test_recovery_logic(&self) -> () {
        // Identify Llama Server process, KILL IT, and verify it comes back.
        // TODO: import time
        // TODO: from utils import is_port_active
        assert!(is_port_active(config::llm_port), "Engine not running at start of test");
        let mut target_pid = None;
        for p in psutil.process_iter(vec!["pid".to_string(), "name".to_string(), "cmdline".to_string()]).iter() {
            if !p.info["name".to_string()].to_lowercase().contains(&"llama-server".to_string()) {
                continue;
            }
            let mut target_pid = p.info["pid".to_string()];
            break;
        }
        if !target_pid {
            pytest.skip("Could not find llama-server process to kill".to_string());
        }
        println!("\n💀 KILLING PID {} FOR SCIENCE...", target_pid);
        psutil.Process(target_pid).kill();
        let mut recovered = false;
        for _ in 0..15.iter() {
            std::thread::sleep(std::time::Duration::from_secs_f64(1));
            if is_port_active(config::llm_port) {
                let mut recovered = true;
                break;
            }
        }
        assert!(recovered, "Orchestrator FAILED to restart the engine after crash!");
        println!("{}", "\n✅ Engine successfully resurrected! 🧟".to_string());
    }
}
